### PR TITLE
Updated bdcn_loss2 to match TF's one

### PR DESCRIPTION
**Pull request by AdrienCourtois**
The discrepancy between TF's loss and PyTorch's one made the training impossible in PyTorch. With this fix, the loss decreases correctly. 
I verified the implementation by running the following test:

```python
pred = [tf.random.normal([8,500,500,1], stddev=2.0) for _ in range(7)]
y = tf.random.normal([8,500,500,1], mean=0.5, stddev=1.0)
y = tf.where(y >= 0, 1., 0.)

preds, loss = pre_process_binary_cross_entropy(losses.BinaryCrossentropy(), pred, y)
print("TensorFlow loss", loss)

pred = [torch.Tensor(x.numpy()).permute(0,3,1,2) for x in pred]
y = torch.Tensor(y.numpy()).permute(0,3,1,2)

loss = sum([bdcn_loss2(preds, y, l_w) for preds, l_w in zip(pred, np.ones(len(pred)))])
print("PyTorch loss", loss)
```
The output is the following:
```
TensorFlow loss tf.Tensor(4.2836795, shape=(), dtype=float32)
PyTorch loss tensor(4.2837)
```

### CODE REVIEW

**Comment by xavysp** [comment]
Hi, thanks for your help, let me train again and see the quantitative difference then I'm update.

**Comment by AdrienCourtois** [comment]
Hello,
Please note any difference in the final performance could be attributed to one of the following:

- Not the same number of epochs between TF and PyTorch
- Parameter "beta1" of Adam is set to default in PyTorch (=0.9) while set to 0.5 in TensorFlow
- Parameter "epsilon" of Adam is not the same by default in both frameworks
- There are sensible architecture differences between TF and PyTorch (```self.pre_dense_2``` and ```self.pre_dense_5_0``` are only defined and used in TF)
- The initialization is not the same between TF and PyTorch (ones uses Xavier Uniform while the other uses Xavier Normal)
- The weight decay is not the same in both (in PyTorch it is 1e-4 while it's 1e-3 in TF)
- The "deltas" parameters (ref to the paper) are not the same in both versions. In TensorFlow it's 1. for every delta while in PyTorch those values are different. 

I hope this is helpful! And of course, the job you've done is incredible I'm only pointing out tiny flaws.

**Comment by xavysp** [comment]
Hi @AdrienCourtois 
you are right, basically the DexiNed-Pytorch is for the extender version of Dexined in TF1,  we just want to reduce some unnecessary layers.  But your observations are great If double check again. By the way, I really appreciate your help, please stay tuned for the next update.

Xavier

**Comment by xavysp** [comment]
Hi @AdrienCourtois  I have updates, but I am afraid that  your suggestions  in the loss, it decreases properly,  but in the quantitative comparison  there is a difference of 3% on ODS. Thanks to your loss update, I have re-coded my loss and it is working good, the results are almost the same released back days. check the [loss graph](https://drive.google.com/file/d/1Lo1cpSUb_PO76bK8uS56ik2jGIddzKy8/view?usp=sharing).

**Comment by AdrienCourtois** [comment]
Hello @xavysp,
I must admit this is surprising. There shouldn't be such a difference of scale between TF and PyTorch just so that it works. Plus, it seems the scale only comes from the "reduction" argument, which is exactly equivalent to increasing the learning rate.
Finally, the values of the loss between TF and PyTorch in my commit are equal which suggests the gradient is different. Maybe is there an unwanted ```.detach()``` in the code somewhere?

All in all, I'm glad it works now. Good job!

**Comment by xavysp** [comment]
Hi @AdrienCourtois  and yes it's strange, I'll check the  optimizer. And let you know how it end.

**Comment by xavysp** [comment]
> Hello,
> Please note any difference in the final performance could be attributed to one of the following:
> 
> * Not the same number of epochs between TF and PyTorch
> * Parameter "beta1" of Adam is set to default in PyTorch (=0.9) while set to 0.5 in TensorFlow
> * Parameter "epsilon" of Adam is not the same by default in both frameworks
> * There are sensible architecture differences between TF and PyTorch (`self.pre_dense_2` and `self.pre_dense_5_0` are only defined and used in TF)
> * The initialization is not the same between TF and PyTorch (ones uses Xavier Uniform while the other uses Xavier Normal)
> * The weight decay is not the same in both (in PyTorch it is 1e-4 while it's 1e-3 in TF)
> * The "deltas" parameters (ref to the paper) are not the same in both versions. In TensorFlow it's 1. for every delta while in PyTorch those values are different.
> 
> I hope this is helpful! And of course, the job you've done is incredible I'm only pointing out tiny flaws.


Hi again, since Dexined version in TF2  is not onits final version I have checked the optimizer. And thanks for your observations
I have seen that the ADAM in  TF1 and pytorch, have the same parameters
TF1:  https://www.tensorflow.org/api_docs/python/tf/compat/v1/train/AdamOptimizer
PyTorch:  https://pytorch.org/docs/stable/generated/torch.optim.Adam.html#torch.optim.Adam
 But WD not :(
I'll re-train it 
Cheers,

**Comment by AdrienCourtois** [comment]
Hello @xavysp,
Don't forget about the parameter beta_1 of the Adam optimizer.
- In TF it's manually set to args.beta1  [see here](https://github.com/xavysp/DexiNed/blob/master/DexiNed-TF2/run_model.py#L57) and [here](https://github.com/xavysp/DexiNed/blob/master/DexiNed-TF2/main.py#L62)
- In PyTorch it's not specified, therefore set to the default value beta1=0.9 [see here](https://github.com/xavysp/DexiNed/blob/master/main.py#L396)

Thank you very much,
Adrien

**Comment by xavysp** [comment]
> 
> 
> Hello @xavysp,
> Don't forget about the parameter beta_1 of the Adam optimizer.
> 
>     * In TF it's manually set to args.beta1  [see here](https://github.com/xavysp/DexiNed/blob/master/DexiNed-TF2/run_model.py#L57) and [here](https://github.com/xavysp/DexiNed/blob/master/DexiNed-TF2/main.py#L62)
> 
>     * In PyTorch it's not specified, therefore set to the default value beta1=0.9 [see here](https://github.com/xavysp/DexiNed/blob/master/main.py#L396)
> 
> 
> Thank you very much,
> Adrien

Thanks as soon as I start checking TF2 I'll change it. By the way, I have updated the BDCN_loss2 in DexiNed-Pytorch, quantitatively its the better option.
